### PR TITLE
tests(visual): add Argos CI

### PR DESCRIPTION
**Pull request by oliviertassinari**
### TL;DR

This PR intent to set up the infrastructure needed to **avoid visual regressions**.
It's using [Argos CI](https://www.argos-ci.com) service under the hood. It's 100% free for open source project.

### Problem solved

Argos-CI's mission is to give developers the insight needed to quickly iterating on UIs. Quickly iterating on UIs involves confidence and speed of iteration.
What we have today is a simple screenshots comparison tools that aim to solve the problem of managing those screenshots. You get:
- Smart comparison: choosing the right screenshots to compare against and the diffing logic. 
- Github PR feedback
- Acceptance and rejection action button.

### Why should we care?

The promise is to catch visual regression sooner with less effort. Also, visual regression and functional regression have a positive intersection, for instance, if you have a look at [the first diff of this build](https://www.argos-ci.com/oliviertassinari/instantsearch.js/builds/3690?sample) (Doctolib UI example) you will see a visual regression at the surface but the root cause was a major functional regression: the availabilities system was broken.

It's flexible enough to fit into three categories of regressions use case I have seen:
- Component as input
- functional tests as input (implemented here)
- pages as input

### Result

- Take screenshots during the functional test.
capture d ecran 2017-05-26 a 20 36 48
- Push those screenshots for processing to Argos CI:  [build example](https://www.argos-ci.com/oliviertassinari/SplitMe/builds/3694).
- We have our first notification, only admins can accept the diffs if they think it's expected.
capture d ecran 2017-05-27 a 19 31 35

### API

We are using [webdriverio](https://github.com/webdriverio/webdriverio/) and [wdio-visual-regression-service](https://github.com/zinserjan/wdio-visual-regression-service).
The API is quite simple, we can leverage 3 methods to extend the coverage. It's documented [here](https://github.com/zinserjan/wdio-visual-regression-service#usage).
Alternatively, we could also take screenshots outside of the functional tests.

### CODE REVIEW

File: scripts/test-ci.sh [hunk 1]
```diff
@@ -6,8 +6,11 @@ set -ev # exit when error
 npm test
 # we need to build to run functional tests
 NODE_ENV=production npm run build
+rm -rf functional-tests/screenshots
 INDEX_PAGE=index npm run test:functional
 INDEX_PAGE=index-preact npm run test:functional
+export $ARGOS_TOKEN=eae2c60a241ab5e7d2884bf8f9b397657d70a63a
```
Comment (Haroenv) [inline scripts/test-ci.sh]: should this token not be a travis encrypted value?

Comment (oliviertassinari) [inline scripts/test-ci.sh]: Should be, it's a temporary token.

**Comment by Haroenv** [comment]
Anyway, thanks a lot for doing a pull request! Where can we see the images after it’s been run?

File: functional-tests/config.js [hunk 2]
```diff
@@ -89,7 +118,7 @@ if (process.env.CI === 'true') {
 } else {
   conf = {
     host: '127.0.0.1',
-    port: 24444,
```
Comment (oliviertassinari) [inline functional-tests/config.js]: I'm assuming it's a typo as the default port is `4444`.

File: functional-tests/config.js [hunk 3]
```diff
@@ -1,10 +1,24 @@
+import path from 'path';
+// eslint-disable-next-line import/no-commonjs
+const VisualRegressionCompare = require('wdio-visual-regression-service/compare');
 import testServer from './testServer.js';
 import {clearAll, searchBox} from './utils.js';
-const INDEX_PAGE = process.env.INDEX_PAGE || 'index.html';
+const INDEX_PAGE = process.env.INDEX_PAGE || 'index';
```
Comment (oliviertassinari) [inline functional-tests/config.js]: Fix `.html.html` issue.

**Comment by oliviertassinari** [comment]
@Haroenv It's not finished yet. Looks like async logic of https://github.com/zinserjan/wdio-visual-regression-service/blob/master/src/VisualRegressionLauncher.js#L23 doesn't work with the sync logic here. I'm gonna switch to a simpler integration; [wdio-screenshot](https://github.com/zinserjan/wdio-screenshot).

> Where can we see the images after it’s been run?

A PR status will be posted and point toward something [like this](https://www.argos-ci.com/callemall/material-ui/builds/3176)

**Comment by oliviertassinari** [comment]
@halley124 Here is the first [successful build](https://www.argos-ci.com/oliviertassinari/SplitMe/builds/3694). Now, we gonna need someone with write access to *instantsearch.js* to setup the project. Basically, we need to use a **secret token** instead of the public one here.

**Comment by Haroenv** [comment]
Looks nice @oliviertassinari, we will get back to you on this, thanks for your pr <3

**Comment by vvo** [comment]
@oliviertassinari PM you on twitter to set that up

**Comment by vvo** [comment]
> @halley124

I think you meant @Haroenv, our VP sales is not actively coding :trollface:

**Comment by vvo** [comment]
Also, I dunno how much work that would be but we have a v2 branch that's pretty far from this develop branch. That v2 branch will become the new develop pretty soon.

The functional test system has been a bit updated, have a look: https://github.com/algolia/instantsearch.js/tree/feat/instantsearch.js/v2

**Comment by vvo** [comment]
image
 ? :D

**Comment by oliviertassinari** [comment]
@vvo I came across that configuration, I wanted to be sure 🍭 .
capture d ecran 2017-05-27 a 01 18 34

Sure, I can do the same effort on the v2 branch.

**Comment by vvo** [comment]
> Sure, I can do the same effort on the v2 branch.

That would be awesome and would ease the process for us into moving to Argos.

One thing also that would be good is to get a bit more context on this PR and what Argos aim to solve and how it will be better for InstantSearch projects overall.

We do not use greenkeeper anymore, we might reuse it soon since they started using package locks.

**Comment by oliviertassinari** [comment]
@vvo Still have some stuff to fix to get this PR working. I think that I'm gonna open a new one on the  v2 branch once it's working with a more detail PR summary and close this one.

**Comment by oliviertassinari** [comment]
> One thing also that would be good is to get a bit more context on this PR and what Argos aim to solve.

Argos-CI's mission is to give developers the insight needed to quickly iterating on UIs. Quickly iterating on UIs involves confidence and speed of iteration.
What we have today is a simple screenshots comparison tools that aim to solve the problem of managing those screenshots. You get:
- Smart comparison: choosing the right screenshots to compare against and the diffing logic. 
- Github PR feedback
- Acceptance and rejection action button.

> how it will be better for InstantSearch projects overall

The promise is to catch visual regression sooner with less effort. Also, visual regression and functional regression have a positive intersection, for instance, if you have a look at [the first diff of this build](https://www.argos-ci.com/oliviertassinari/instantsearch.js/builds/3690?sample) you will see a visual regression at the surface but the root cause was a major functional regression: the availabilities system was broken.

It's flexible enough to fit into three categories of regressions use case I have seen:
- Component as input
- functional tests as input (implemented here)
- pages as input

🎉  We have our first notification, I need to fix the flakiness now.

capture d ecran 2017-05-27 a 19 31 35

**Comment by Haroenv** [comment]
What is the way to approve or disapprove of a change, is that once you’re logged in via the UI?

**Comment by oliviertassinari** [comment]
Yes, you need to be logged and have write access to the repository to change the status of a build. I have used the `hide` option to hide the flaky `X ms` information displayed on the screen.

**Comment by oliviertassinari** [comment]
I have realized that you have a [storybook](https://community.algolia.com/react-instantsearch/storybook) that would be a great starting point to add the visual regression tests!
I will work on that next.

**Comment by Haroenv** [comment]
@oliviertassinari, the storybook is in the react instant search [repo](https://github.com/algolia/react-instantsearch/) 😄

**Comment by oliviertassinari** [comment]
Thanks for the link, that's much closer to what we do with [Material UI](https://github.com/callemall/material-ui/tree/next/test/regressions). That don't have to be written in React. is `react-instantsearch` a simple wrapper around component written here? Coming back to the previous categories. For a UI lib `Component as input` provides much more value than the `functional tests as input`. It more suited for end applications.

**Comment by Haroenv** [comment]
It’s a completely separate project

**Comment by oliviertassinari** [comment]
@Haroenv Oh, sounds like there is a variety of needs with instantsearch. That's perfect for bulletproofing Argos CI 😄 . I will focus on nailing the needs here.

**Comment by vvo** [comment]
Awesome!

On InstantSearch.js we have functional tests that you can plug on, also see v2 branch
On React InstantSearch we have a full storybook where you could add Argos, that would be awesome too.

At some point with InstantSearch.js we will maybe share some more parts with React InstantSearch, and that could lead to a separate component lib that would be tested with Argos.

So much fun to come :D

**Comment by bobylito** [comment]
Just checked the results on the argos CI page and I must say that feels like it will push us to do more functional testing 💯

**Comment by oliviertassinari** [comment]
I have rebased the PR on the `feat/instantsearch.js/v2` branch. It's all green now. However, I had to make some small changes as that branch doesn't seem configurated to run on the CI.
Also, I have made some progress with react-instantsearch. I have got 82 test cases extracted from the storybook. I need to stabilise them now.

File: .travis.yml [hunk 4]
```diff
@@ -14,6 +14,7 @@ branches:
   only:
   - master
   - develop
+  - feat/instantsearch.js/v2
```
Comment (vvo) [inline .travis.yml]: 🆗

**Comment by vvo** [comment]
> as that branch doesn't seem configurated to run on the CI.

Seems like it's not yet indeed, thanks for the change.

File: functional-tests/config.js [hunk 5]
```diff
@@ -1,7 +1,18 @@
+import path from 'path';
+// eslint-disable-next-line import/no-commonjs
+const VisualRegressionCompare = require('wdio-visual-regression-service/compare');
```
Comment (vvo) [inline functional-tests/config.js]: Can you use import syntax or is it not working?

File: .gitignore [hunk 6]
```diff
@@ -5,3 +5,6 @@ npm-debug.log
 .DS_Store
 docs/
 !scripts/docs/
+
+# Ignore Argos CI screenshots.
+functional-tests/screenshots/
```
Comment (vvo) [inline .gitignore]: Can it be only dealt with by travis, i.e. what's the use of git keeping this folder in dev env?

File: scripts/test-ci.sh [hunk 7]
```diff
@@ -7,11 +7,10 @@ npm test
 
 # we need to build to run functional tests
 NODE_ENV=production npm run build
+rm -rf functional-tests/screenshots
```
Comment (vvo) [inline scripts/test-ci.sh]: Maybe we could also ensure the folder exists here or just configure some module to create it automatically and empty it.

File: scripts/test-ci.sh [hunk 8]
```diff
@@ -7,11 +7,10 @@ npm test
...
 INDEX_PAGE=index npm run test:functional
 ./scripts/validate-commit-msgs.sh
-
-# check we are able to generate docs
-npm run docs:jsdoc
+npm run argos
```
Comment (vvo) [inline scripts/test-ci.sh]: Can you add a comment that says what is argos briefly

File: functional-tests/config.js (hunk 5 above)
Comment (oliviertassinari) [inline functional-tests/config.js]: Turn out we can, replacing it.

File: .gitignore (hunk 6 above)
Comment (oliviertassinari) [inline .gitignore]: It's here in case someone is running the e2e tests on his development environment. We don't want the screenshots to be commited.

File: scripts/test-ci.sh (hunk 7 above)
Comment (oliviertassinari) [inline scripts/test-ci.sh]: No need for that line. I'm removing it.

Comment (oliviertassinari) [inline scripts/test-ci.sh]: The folder is already automatically created by `wdio-visual-regression-service`.

**Comment by vvo** [comment]
LGTM I'll let @bobylito decide when to merge this

**Comment by bobylito** [comment]
This v2 is gonna be packed with devtools :D (we just added a story book like tool, thanks to @iam4x)

**Comment by oliviertassinari** [comment]
🎉 thanks guys.
